pnpm publish exits with code 0, even if it fails

### pnpm version: 7.18.1

### Code to reproduce the issue: run `pnpm publish` without / with wrong access token, incorrect registry, etc.

### Expected behavior: pnpm exits with code != 0

### Actual behavior: pnpm exits with code 0

### Additional information:

 - `node -v` prints: 16.18.1
 - Windows, macOS, or Linux?: ubuntu linux (github)

I tried `pnpm publish` with incorrect registry it seems exit code to be 1:

```
$ pnpm publish --registry='https://xfafaf.com'
npm notice 
npm notice 📦  await-o-test-alpha-test@1.2.0-alpha.1
npm notice === Tarball Contents === 
npm notice 309B package.json
npm notice 0B   test.js     
npm notice === Tarball Details === 
npm notice name:          await-o-test-alpha-test                  
npm notice version:       1.2.0-alpha.1                            
npm notice filename:      await-o-test-alpha-test-1.2.0-alpha.1.tgz
npm notice package size:  304 B                                    
npm notice unpacked size: 309 B                                    
npm notice shasum:        54c316645f96ddb83a81248deead425903c7e89e 
npm notice integrity:     sha512-QKS1WXo36ZMFL[...]n3iUQOFdOPiYg== 
npm notice total files:   2                                        
npm notice 
npm ERR! code ENEEDAUTH
npm ERR! need auth This command requires you to be logged in to https://xfafaf.com/
npm ERR! need auth You need to authorize this machine using `npm adduser`

$ echo $?
1
```

--- a/.changeset/quick-worms-rush.md
+++ b/.changeset/quick-worms-rush.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/plugin-commands-publishing": patch
+"pnpm": patch
+---
+
+When publish some package throws an error, the exit code should be non-zero [#5528](https://github.com/pnpm/pnpm/issues/5528).

--- a/releasing/plugin-commands-publishing/src/publish.ts
+++ b/releasing/plugin-commands-publishing/src/publish.ts
@@ -174,12 +174,12 @@ Do you want to continue?`,
     }
   }
   if (opts.recursive && (opts.selectedProjectsGraph != null)) {
-    await recursivePublish({
+    const { exitCode } = await recursivePublish({
       ...opts,
       selectedProjectsGraph: opts.selectedProjectsGraph,
       workspaceDir: opts.workspaceDir ?? process.cwd(),
     })
-    return
+    return { exitCode }
   }
   if ((params.length > 0) && params[0].endsWith('.tgz')) {
     const { status } = runNpm(opts.npmPath, ['publish', ...params])

--- a/releasing/plugin-commands-publishing/src/recursivePublish.ts
+++ b/releasing/plugin-commands-publishing/src/recursivePublish.ts
@@ -56,7 +56,7 @@ Partial<Pick<Config,
 
 export async function recursivePublish (
   opts: PublishRecursiveOpts & Required<Pick<Config, 'selectedProjectsGraph'>>
-) {
+): Promise<{ exitCode: number }> {
   const pkgs = Object.values(opts.selectedProjectsGraph).map((wsPkg) => wsPkg.package)
   const resolve = createResolver({
     ...opts,
@@ -82,7 +82,6 @@ export async function recursivePublish (
   })
   const publishedPkgDirs = new Set(pkgsToPublish.map(({ dir }) => dir))
   const publishedPackages = []
-  let exitCode = 0
   if (publishedPkgDirs.size === 0) {
     logger.info({
       message: 'There are no new packages that should be published',
@@ -123,8 +122,8 @@ export async function recursivePublish (
         }, [pkg.dir])
         if (publishResult?.manifest != null) {
           publishedPackages.push(pick(['name', 'version'], publishResult.manifest))
-        } else if (exitCode === 0 && publishResult?.exitCode) {
-          exitCode = publishResult.exitCode
+        } else if (publishResult?.exitCode) {
+          return { exitCode: publishResult.exitCode }
         }
       }
     }
@@ -132,7 +131,7 @@ export async function recursivePublish (
   if (opts.reportSummary) {
     await writeJsonFile(path.join(opts.lockfileDir ?? opts.dir, 'pnpm-publish-summary.json'), { publishedPackages })
   }
-  return { exitCode }
+  return { exitCode: 0 }
 }
 
 async function isAlreadyPublished (

--- a/releasing/plugin-commands-publishing/src/recursivePublish.ts
+++ b/releasing/plugin-commands-publishing/src/recursivePublish.ts
@@ -82,6 +82,7 @@ export async function recursivePublish (
   })
   const publishedPkgDirs = new Set(pkgsToPublish.map(({ dir }) => dir))
   const publishedPackages = []
+  let exitCode = 0
   if (publishedPkgDirs.size === 0) {
     logger.info({
       message: 'There are no new packages that should be published',
@@ -122,6 +123,8 @@ export async function recursivePublish (
         }, [pkg.dir])
         if (publishResult?.manifest != null) {
           publishedPackages.push(pick(['name', 'version'], publishResult.manifest))
+        } else if (exitCode === 0 && publishResult?.exitCode) {
+          exitCode = publishResult.exitCode
         }
       }
     }
@@ -129,6 +132,7 @@ export async function recursivePublish (
   if (opts.reportSummary) {
     await writeJsonFile(path.join(opts.lockfileDir ?? opts.dir, 'pnpm-publish-summary.json'), { publishedPackages })
   }
+  return { exitCode }
 }
 
 async function isAlreadyPublished (

--- a/releasing/plugin-commands-publishing/test/recursivePublish.ts
+++ b/releasing/plugin-commands-publishing/test/recursivePublish.ts
@@ -284,3 +284,29 @@ test('recursive publish writes publish summary', async () => {
     })
   }
 })
+
+test('when publish some package throws an error, exit code should be non-zero', async () => {
+  preparePackages([
+    {
+      name: '@pnpmtest/test-recursive-publish-project-5',
+      version: '1.0.0',
+    },
+    {
+      name: '@pnpmtest/test-recursive-publish-project-6',
+      version: '1.0.0',
+    },
+  ])
+
+  // Throw ENEEDAUTH error when publish.
+  await fs.writeFile('.npmrc', 'registry=https://__fake_npm_registry__.com', 'utf8')
+
+  const result = await publish.handler({
+    ...DEFAULT_OPTS,
+    ...await readProjects(process.cwd(), []),
+    dir: process.cwd(),
+    recursive: true,
+    force: true,
+  }, [])
+
+  expect(result?.exitCode).toBe(1)
+})

